src/storage/transaction/TossEdgeIterator.h

```diff
@@ -0,0 +1,227 @@
+                    auto tryLockData = recoverEdges_.back()->tryWLock();
+                    if (tryLockData && tryLockData->first.empty()) {
+                        tryLockData->first = iter_->key().str();
+                        tryLockData->second = iter_->val().str();
+                    }
+                    lastIsLock_ = false;
+                }
+                continue;
+            } else if (NebulaKeyUtils::isLock(planContext_->vIdLen_, iter_->key())) {
+                // LOG_IF(INFO, FLAGS_trace_toss) << "next to a lock, ver=" << ver;
+                std::string rawKey = NebulaKeyUtils::edgeKey(iter_->key());
+                auto rank = NebulaKeyUtils::getRank(planContext_->vIdLen_, rawKey);
+                auto dstId = NebulaKeyUtils::getDstId(planContext_->vIdLen_, rawKey).str();
+                if (!lastIsLock_ && rank == lastRank_ && dstId == lastDstId_) {
+                    continue;
+                }
+                if (rank != lastRank_ || dstId != lastDstId_) {
+                    recoverEdges_.emplace_back(
+                        std::make_shared<folly::Synchronized<KV>>(std::make_pair("", "")));
+                }
+
+                resumeTasks_.emplace_back(
+                    planContext_->env_->txnMan_->resumeTransaction(planContext_->vIdLen_,
+                                                                   planContext_->spaceId_,
+                                                                   iter_->key().str(),
+                                                                   iter_->val().str(),
+                                                                   recoverEdges_.back()));
+                lastRank_ = NebulaKeyUtils::getRank(planContext_->vIdLen_, rawKey);
+                lastDstId_ = NebulaKeyUtils::getDstId(planContext_->vIdLen_, rawKey).str();
+                lastIsLock_ = true;
+                continue;
+            } else {
+                LOG_IF(INFO, FLAGS_trace_toss)
+                    << "next to a weird record: \n"
+                    << folly::hexDump(iter_->key().data(), iter_->key().size())
+                    << folly::hexDump(iter_->val().data(), iter_->val().size());
+            }
+        }
+        LOG_IF(INFO, FLAGS_trace_toss) << "no more iter to read";
+
+        if (needWaitResumeTask_) {
+            folly::collectAll(resumeTasks_).wait();
+            needWaitResumeTask_ = false;
+            recoverEdgesIter_ = recoverEdges_.begin();
+            firstLoop_ = true;
+        }
+
+        while (recoverEdgesIter_ != recoverEdges_.end()) {
+            if (recoverEdgesIter_ != recoverEdges_.begin() || firstLoop_ != true) {
+                recoverEdgesIter_++;
+            }
+            if (recoverEdgesIter_ == recoverEdges_.end()) {
+                return;
+            }
+            firstLoop_ = false;
+            auto code = (*recoverEdgesIter_)->withRLock([&](const auto& data) {
```

Comment:
After a rethink, do we need lock for `recoverEdges`? When `iter_` is valid, we will scan over all possible lock and record. When `iter_` hit the end, we will `collectAll` in line 153, there should not be any conflicts? Am I right?

Reply:
make sense.